### PR TITLE
Upgrade Gradle to 8.2.1 and AGP to 8.1.0

### PR DESCRIPTION
**Pull request by oguzkocer**
## Description
This PR upgrades Gradle to `8.2.1` & Android Gradle Plugin to `8.0.2`.

* f741825decb322e683af8d7e1743e75e10e91938 upgrades Gradle to `8.2.1`. The changes are generated by running `./gradlew wrapper --gradle-version 8.2.1 --distribution-type all`. This is how we've updated all our other projects, but [as requested](https://github.com/Automattic/pocket-casts-android/pull/1208#discussion_r1274240958) by @geekygecko, I've also added the `distributionSha256Sum` in fffb8ab9a83635b4740c227948c1a31e928ce105 by running `./gradlew wrapper --gradle-version 8.2.1 --distribution-type all --gradle-distribution-sha256-sum 7c3ad722e9b0ce8205b91560fd6ce8296ac3eadf065672242fd73c06b8eeb6ee` and using the checksum from [this page](https://gradle.org/release-checksums/). Gradle Wrapper 8 already includes a new property `validateDistributionUrl`, so I am not sure if the checksum is still necessary, but it's easy enough to keep it.
* 1cc436ce5739c76303dc9c02fd87f574a94fe06b adds `.java-version` file which is how our CI will know to use JDK 17 when building this project. Using JDK 17 is a requirement for upgrading Android Gradle Plugin to `8.0` as it'll not build otherwise.
* 11a9d66b167f757d5b81ce8e69d4547785a645d8 enables `buildConfig` AGP feature for all modules. Without this, the build will fail with `defaultConfig contains custom BuildConfig fields, but the feature is disabled.`. I initially tried only adding it to the modules that require it, but that felt pretty endless and I figured if we are going to add this to most modules, we might as well add it to all modules to be consistent. For this particular project, one possibility might be to have a separate module just for the `BuildConfig` values, so the rest of the modules don't get involved with it. However, that'd be outside the scope of this PR and it's more of a product team decision rather than an infrastructure one.
* 46eb324b693ae2119b1dc34f1ae1073c70ac6c1b updates AGP to `8.0.2`.
* 0696ef613174ab2c7cf77750cb33c4fbb435a3a1 resolves missing `R` class errors in AGP `8.0.2` by adding `dontwarn` proguard rules as suggested by AGP build error. These errors are already [present](https://buildkite.com/automattic/pocket-casts-android/builds/3185#01898b61-a791-4739-af03-1e137e5a1e1f/323-1185) in current `main` for `release` variant, although in AGP `7+` they are just warnings:

```
▸ WARNING:Missing classes detected while running R8. Please add the missing classes or apply additional keep rules that are generated in /var/lib/buildkite-agent/builds/ci-android-i-06c818c993131120b-1/automattic/pocket-casts-android/modules/services/model/build/outputs/mapping/release/missing_rules.txt.
▸ WARNING:R8: Missing class au.com.shiftyjelly.pocketcasts.localization.R$string (referenced from: void au.com.shiftyjelly.pocketcasts.models.type.SubscriptionFrequency.<clinit>())
```

I am far from confident that this is the best way to address the proguard issues as my experience with it is almost non-existent. I think this resolution will keep the existing behavior which I guess works, but I would have expected some kind of `keep class` instead. However, I couldn't get any other configuration to work. @geekygecko maybe you can help with this?

## Testing Instructions

Smoke test `app`, `auto` & `wear`

## Checklist
- [x] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [x] I have considered whether it makes sense to add tests for my changes
- [x] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [x] Any jetpack compose components I added or changed are covered by compose previews
 
#### I have tested any UI changes...

- [x] with different themes
- [x] with a landscape orientation
- [x] with the device set to have a large display and font size
- [x] for accessibility with TalkBack

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties
Comment (geekygecko) [inline gradle/wrapper/gradle-wrapper.properties:4]: Sorry I know this is a draft but would it be good to keep the `distributionSha256Sum`

Comment (oguzkocer) [reply gradle/wrapper/gradle-wrapper.properties:7]: Addressed in https://github.com/Automattic/pocket-casts-android/pull/1208/commits/fffb8ab9a83635b4740c227948c1a31e928ce105 by using the release checksum from [this page](https://gradle.org/release-checksums/).

**Comment by oguzkocer** [comment]
@geekygecko @ParaskP7 This PR is ready for review, but let's hold off on merging it for now. We'll soon publish an announcement for all Android developers about these AGP updates and JDK 17 in which we'll include instructions for how to update our local environments and what to watch out for after these PRs are merged. That is why I am keeping the PR as draft for now.

**Review by ParaskP7 — APPROVED** [review]
👋 @oguzkocer !

I have reviewed and tested this PR as per the instructions, everything works as expected, great job! 🌟 🌟 🌟

-----

I am just unsure about the proguard configuration, but I am going to tentatively approve this PR, just to prevent from blocking it.

> https://github.com/Automattic/pocket-casts-android/commit/0696ef613174ab2c7cf77750cb33c4fbb435a3a1 resolves missing R class errors in AGP 8.0.2 by adding dontwarn proguard rules as suggested by AGP build error. These errors are already [present](https://buildkite.com/automattic/pocket-casts-android/builds/3185#01898b61-a791-4739-af03-1e137e5a1e1f/323-1185) in current main for release variant, although in AGP 7+ they are just warnings:
> 
> ...
> 
> I am far from confident that this is the best way to address the proguard issues as my experience with it is almost non-existent. I think this resolution will keep the existing behavior which I guess works, but I would have expected some kind of keep class instead. However, I couldn't get any other configuration to work.

👍 and same here... 😊 

> @geekygecko maybe you can help with this?

🙏

**Comment by oguzkocer** [comment]
@geekygecko As I mentioned I am not experienced with Proguard at all. So, can I leave that bit to you?

Please feel free to directly push any changes you'd like to this branch.

**Comment by ParaskP7** [comment]
👋 @oguzkocer !

I am trying to figure out why CI is failing, maybe because of [this](https://github.com/Automattic/buildkite-ci/pull/236) (similar to [DOAndroid](https://github.com/bloom/DayOne-Android/pull/2526)), but when I try to see the [build](https://buildkite.com/automattic/pocket-casts-android/builds/3226) in order to trigger a rerun, I get a page not found. 🤔

FYI: The only 2 [Pocket Casts](https://buildkite.com/automattic?tags%5B%5D=%3Aheadphones%3A%20Pocket%20Casts) projects I am seeing in Buildkite are one for iOS and one for Android, but pointing to [Pocket Casts Android Old](https://buildkite.com/automattic/pocket-casts-android-old). Do we have a new one? ❓

**Comment by oguzkocer** [comment]
@ParaskP7 I had to update our AMI yesterday to install Android Build Tools `33.0.1`. I just forgot to retry the CI jobs for this repository.

**Comment by ParaskP7** [comment]
> @ParaskP7 I had to update our AMI yesterday to install Android Build Tools 33.0.1. I just forgot to retry the CI jobs for this repository.

That's right @oguzkocer and thanks (I am actually aware of that). 🥇 

I was just saying that I couldn't trigger the retry myself because I get a [page not found](https://buildkite.com/automattic/pocket-casts-android/builds/3226). Am I missing something, a permission of some kind for Buildkite on this project? 🤔

**Comment by oguzkocer** [comment]
> That's right @oguzkocer and thanks (I am actually aware of that). 🥇

You even linked the relevant PR, sorry 🤦 

> I was just saying that I couldn't trigger the retry myself because I get a [page not found](https://buildkite.com/automattic/pocket-casts-android/builds/3226). Am I missing something, a permission of some kind for Buildkite on this project? 🤔

@ParaskP7 I added you as a maintainer to the Buildkite team, so that should work now.

**Comment by oguzkocer** [comment]
@geekygecko This update has been merged for all the other projects, but since I didn't get your approval, I didn't want to merge this PR. Please feel free to merge this yourself or let me know if you have any feedback you'd like to be addressed before the merge.

**Comment by trietbui85** [comment]
>https://github.com/Automattic/pocket-casts-android/commit/1cc436ce5739c76303dc9c02fd87f574a94fe06b adds .java-version file which is how our CI will know to use JDK 17 when building this project. Using JDK 17 is a requirement for upgrading Android Gradle Plugin to 8.0 as it'll not build otherwise.

Hi @oguzkocer , I don't know about that `.java-version`. Where to get more info about it?

File: modules/features/account/build.gradle
Comment (trietbui85) [inline modules/features/account/build.gradle:9]: I'm not sure if we need to access `BuildConfig` in feature module? If not, it should be disabled.

Comment (oguzkocer) [reply modules/features/account/build.gradle:9]:
```
* What went wrong:
A problem occurred configuring project ':modules:features:account'.
> defaultConfig contains custom BuildConfig fields, but the feature is disabled.
  To enable the feature, add the following to your module-level build.gradle:
  `android.buildFeatures.buildConfig true`
```

Yes it's necessary for this module.

**Comment by oguzkocer** [comment]
> >https://github.com/Automattic/pocket-casts-android/commit/1cc436ce5739c76303dc9c02fd87f574a94fe06b adds .java-version file which is how our CI will know to use JDK 17 when building this project. Using JDK 17 is a requirement for upgrading Android Gradle Plugin to 8.0 as it'll not build otherwise.
> 
> Hi @oguzkocer , I don't know about that `.java-version`. Where to get more info about it?

@trietbui85 The initial explanation pretty much covers why `.java-version` file is necessary. If you are asking about how CI uses this file, we use [`jenv`](https://www.jenv.be/) which selects to correct JDK per project.

**Comment by geekygecko** [comment]
@oguzkocer Today I will test the release version on mobile, Wear OS and Automotive. If everything goes well, I will merge the PR.

**Comment by oguzkocer** [comment]
Thanks @geekygecko! Let me know if you need anything.

**Review by geekygecko — APPROVED** [review]

**Comment by ParaskP7** [comment]
> You even linked the relevant PR, sorry 🤦

😊 

> @ParaskP7 I added you as a maintainer to the Buildkite team, so that should work now.

Awesome @oguzkocer , thank you so much for adding me as a maintainer, I can now verify that I have access to this project in Buildkite. ✅

**Pull request merged**
